### PR TITLE
Documentation 

### PR DESCRIPTION
**Pull request by JeanWi**
Documentation finished of
- Technology class
- Sink class
- Conv3

### CODE REVIEW

File: src/components/technologies/genericTechnologies/stor.py [hunk 1]
```diff
@@ -410,7 +449,15 @@ def init_energyconsumption_out(const, t, car):
...
 
-    def _define_stor_capex(self, b_tec, data):
+    def _define_stor_capex(self, b_tec: Block, data: dict) -> Block:
+        """
+        Construct constraints for the storage CAPEX
+
+        :param b_tec: technology Block
+        :param dict data: input data
+        :return: technology Block with the constraints for the stor technology
+        :rtype: b_tec
```
Comment (IngeOssentjuk) [inline src/components/technologies/genericTechnologies/stor.py:459]: type is Block

File: src/components/technologies/genericTechnologies/stor.py [hunk 2]
```diff
@@ -84,11 +94,13 @@ def __init__(self, tec_data):
         self.fitted_performance = FittedPerformance()
         self.flexibility_data = tec_data["Flexibility"]
 
-    def fit_technology_performance(self, climate_data, location):
+    def fit_technology_performance(self, climate_data: pd.DataFrame, location):
```
Comment (IngeOssentjuk) [inline src/components/technologies/genericTechnologies/stor.py:97]: location: dict

File: src/components/technologies/genericTechnologies/stor.py [hunk 3]
```diff
@@ -155,13 +167,18 @@ def fit_technology_performance(self, climate_data, location):
...
 
-        :param b_tec:
-        :param energyhub:
-        :return: b_tec
+        :param b_tec: technology Block
+        :param dict data: input data
+        :param Set set_t: set of timesteps in the model
+        :param Set set_t_clustered: set of timesteps when clustering algorithm is used
+        :return: technology Block with the constraints for the storage technology
+        :rtype: b_tec
```
Comment (IngeOssentjuk) [inline src/components/technologies/genericTechnologies/stor.py:181]: rtype is Block

File: src/components/technologies/genericTechnologies/stor.py [hunk 4]
```diff
@@ -498,12 +548,12 @@ def _define_stor_capex(self, b_tec, data):
...
 
-    def write_results_tec_operation(self, h5_group, model_block):
+    def write_results_tec_operation(self, h5_group: h5py.Group, model_block: Block):
         """
         Function to report results of technologies after optimization
 
-        :param b_tec: technology model block
-        :return: dict results: holds results
+        :param Block b_tec: technology model block
+        :param h5py.Group h5_group: technology model block
```
Comment (IngeOssentjuk) [inline src/components/technologies/genericTechnologies/stor.py:556]: group explanation: group in HDF5 file of model run to which these results are saved.

File: src/components/technologies/genericTechnologies/stor.py [hunk 5]
```diff
@@ -534,12 +589,12 @@ def write_results_tec_design(self, h5_group, model_block):
...
 
-    def _define_ramping_rates(self, b_tec):
+    def _define_ramping_rates(self, b_tec: Block):
         """
         Constraints the inputs for a ramping rate. Implemented for input and output
 
-        :param b_tec: technology model block
-        :return:
+        :param Block b_tec: technology model block
+        :return: technology Block with the constraints for the dynamic behaviour
```
Comment (IngeOssentjuk) [inline src/components/technologies/genericTechnologies/stor.py:597]: add :rtype: Block

File: src/components/technologies/genericTechnologies/sink.py [hunk 6]
```diff
@@ -70,7 +71,7 @@ def __init__(self, tec_data: dict):
         self.fitted_performance = FittedPerformance()
         self.flexibility_data = tec_data["Flexibility"]
 
-    def fit_technology_performance(self, climate_data: pd.DataFrame, location: dict):
+    def fit_technology_performance(self, climate_data: pd.DataFrame, location):
```
Comment (IngeOssentjuk) [inline src/components/technologies/genericTechnologies/sink.py:74]: location: dict

File: src/components/technologies/technology.py [hunk 7]
```diff
@@ -27,6 +28,203 @@ class Technology(ModelComponent):
...
     constraints and sets of a technology.
+
+    This function is extented in the generic/specific technology classes. It adds Sets, Parameters, Variables and
+    Constraints that are common for all technologies
+    The following description is true for new technologies. For existing technologies a few adaptions are made
+    (see below).
+    When CCS is available, we add heat and electricity to the input carriers Set and CO2captured to the output
+    carriers Set. Moreover, we create extra Parameters and Variables equivalent to the ones created for the
+    technology, but specific for CCS. In addition, we create Variables that are the sum of the input, output,
+    CAPEX and OPEX of the technology and of CCS. We calculate the emissions of the techology discounting already
```
Comment (IngeOssentjuk) [inline src/components/technologies/technology.py:39]: typo: technology

File: src/components/technologies/technology.py [hunk 8]
```diff
@@ -347,6 +442,13 @@ def init_aggregate_output(const, t, car):
         return b_tec
 
     def _aggregate_cost(self, b_tec):
```
Comment (IngeOssentjuk) [inline src/components/technologies/technology.py:444]: in the  function b_tec: Block

File: src/components/technologies/technology.py [hunk 9]
```diff
@@ -390,10 +495,9 @@ def init_aggregate_opex_fixed(const):
 
     def write_results_tec_design(self, h5_group, model_block):
```
Comment (IngeOssentjuk) [inline src/components/technologies/technology.py:496]: in the method, h5_group: h5py.Group, model_block: Block

File: src/components/technologies/technology.py [hunk 10]
```diff
@@ -428,7 +532,11 @@ def write_results_tec_design(self, h5_group, model_block):
             )
 
     def write_results_tec_operation(self, h5_group, model_block):
```
Comment (IngeOssentjuk) [inline src/components/technologies/technology.py:534]: same

File: src/components/technologies/technology.py [hunk 11]
```diff
@@ -508,9 +615,13 @@ def scale_model(self, b_tec, model, config):
 
         return model
 
-    def _define_input_carriers(self, b_tec):
+    def _define_input_carriers(self, b_tec: Block):
```
Comment (IngeOssentjuk) [inline src/components/technologies/technology.py:618]: add -> Block

File: src/components/technologies/technology.py [hunk 12]
```diff
@@ -532,9 +643,13 @@ def _define_input_carriers(self, b_tec):
 
         return b_tec
 
-    def _define_output_carriers(self, b_tec):
+    def _define_output_carriers(self, b_tec: Block):
```
Comment (IngeOssentjuk) [inline src/components/technologies/technology.py:646]: add return

File: src/components/technologies/technology.py [hunk 13]
```diff
@@ -604,13 +715,9 @@ def _define_capex(self, b_tec, data):
         """
```
Comment (IngeOssentjuk) [inline src/components/technologies/technology.py:715]: add :param: data to docstring and -> Block to method

File: src/components/technologies/technology.py [hunk 14]
```diff
@@ -754,15 +864,13 @@ def _define_input(self, b_tec, data):
...
```
Comment (IngeOssentjuk) [inline src/components/technologies/technology.py:864]: add: datatypes in method and data to docstring param

File: src/components/technologies/technology.py [hunk 15]
```diff
@@ -811,13 +919,13 @@ def _define_output(self, b_tec, data):
...
```
Comment (IngeOssentjuk) [inline src/components/technologies/technology.py:919]: same

File: src/components/technologies/technology.py [hunk 16]
```diff
@@ -864,6 +972,10 @@ def init_output_bounds(bounds, t, car):
     def _define_opex(self, b_tec, data):
```
Comment (IngeOssentjuk) [inline src/components/technologies/technology.py:972]: same

File: src/components/technologies/technology.py [hunk 17]
```diff
@@ -905,6 +1018,10 @@ def init_opex_variable(const, t):
     def _define_emissions(self, b_tec, data):
```
Comment (IngeOssentjuk) [inline src/components/technologies/technology.py:1018]: same

File: src/components/technologies/technology.py [hunk 18]
```diff
@@ -1002,6 +1120,10 @@ def init_tec_emissions_neg(const, t):
     def _define_ccs_performance(self, b_tec, data):
```
Comment (IngeOssentjuk) [inline src/components/technologies/technology.py:1120]: same

File: src/components/technologies/technology.py [hunk 19]
```diff
@@ -1105,10 +1225,13 @@ def init_input_ccs(const, t, car):
     def _define_ccs_emissions(self, b_tec, data):
```
Comment (IngeOssentjuk) [inline src/components/technologies/technology.py:1225]: same

File: src/components/technologies/technology.py [hunk 20]
```diff
@@ -1172,6 +1295,13 @@ def init_tec_emissions_neg(const, t):
         return b_tec
 
     def _define_ccs_costs(self, b_tec, data):
```
Comment (IngeOssentjuk) [inline src/components/technologies/technology.py:1297]: same

File: src/components/technologies/technology.py [hunk 21]
```diff
@@ -1296,6 +1413,10 @@ def init_opex_variable_ccs(const, t):
     def _define_auxiliary_vars(self, b_tec, data):
```
Comment (IngeOssentjuk) [inline src/components/technologies/technology.py:1413]: same

**Review by IngeOssentjuk — APPROVED** [review]
minor additions required

**Review by lucabert01 — COMMENTED** [review]
Looks good, thanks Jan! One thing, is the constraint documentation for the SINK removed on purpose?

**Comment by JeanWi** [comment]
@lucabert01: It is not, is it? its all still there but just moved to the docstring of the class?

**Review by lucabert01 — APPROVED** [review]

**Review by lucabert01 — APPROVED** [review]

**Review by julia1071 — APPROVED** [review]

**Pull request merged**